flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkResourceExceptionUtils.java

```diff
@@ -0,0 +1,118 @@
+        } catch (Exception e) {
+            // Rollback to setting error string/message to CRD
+            ((AbstractFlinkResource<?, ?>) resource)
+                    .getStatus()
+                    .setError(
+                            (e instanceof ReconciliationException)
+                                    ? e.getCause().toString()
+                                    : e.toString());
+        }
+    }
+
+    private static <R extends AbstractFlinkResource> boolean getStackTraceEnabled(R resource) {
+        return Optional.ofNullable((AbstractFlinkSpec) resource.getSpec())
+                .map(AbstractFlinkSpec::isStackTraceEnabled)
+                .orElse(false);
+    }
+
+    private static FlinkResourceException getFlinkResourceException(
+            Throwable throwable, boolean isStackTraceEnabled) {
+        FlinkResourceException flinkResourceException =
+                convertToFlinkResourceException(throwable, isStackTraceEnabled);
+
+        flinkResourceException.setThrowableList(
+                ExceptionUtils.getThrowableList(
+                                throwable.getCause())
+                        .stream()
+                        .map((t) -> convertToFlinkResourceException(t, false))
+                        .collect(Collectors.toList()));
+
+        return flinkResourceException;
+    }
+
+    private static FlinkResourceException convertToFlinkResourceException(
+            Throwable throwable, boolean isStackTraceEnabled) {
+        FlinkResourceException flinkResourceException =
+                FlinkResourceException.builder()
+                        .type(throwable.getClass().getName())
+                        .message(throwable.getMessage())
+                        .build();
+
+        if (isStackTraceEnabled) {
+            flinkResourceException.setStackTraceElements(throwable.getStackTrace());
+        }
+
+        enrichMetadata(throwable, flinkResourceException);
+
+        return flinkResourceException;
+    }
+
+    private static void enrichMetadata(
+            Throwable throwable, FlinkResourceException flinkResourceException) {
+        if (throwable instanceof RestClientException) {
+            flinkResourceException.setAdditionalMetadata(
+                    Map.of(
+                            "httpResponseCode",
+                            ((RestClientException) throwable).getHttpResponseStatus().code()));
+        }
+
+        if (throwable instanceof DeploymentFailedException) {
+            flinkResourceException.setAdditionalMetadata(
+                    Map.of("reason", ((DeploymentFailedException) throwable).getReason()));
+        }
+
+        // This section can be extended to enrich more metadata in the future.
+    }
+
+    private static String convertToJson(FlinkResourceException flinkResourceException)
+            throws JsonProcessingException {
+        return new ObjectMapper().writeValueAsString(flinkResourceException);
```

Comment:
We should re-use a `private static final ObjectMapper`

Reply:
Updated, thanks